crates/matrix-sdk/src/test_utils/mocks.rs

```diff
@@ -801,10 +868,286 @@ impl<'a> MockEndpoint<'a, RoomSendEndpoint> {
+    ///     }))
+    ///     .ok(event_id)
+    ///     .expect(1)
+    ///     .mount()
+    ///     .await;
+    ///
+    /// let mut content = RoomPowerLevelsEventContent::new();
+    /// // Update the power level to a non default value.
+    /// // Otherwise it will be skipped from serialization.
+    /// content.redact = 51.into();
+    ///
+    /// let response = room.send_state_event(content).await?;
+    ///
+    /// assert_eq!(
+    ///     event_id,
+    ///     response.event_id,
+    ///     "The event ID we mocked should match the one we received when we sent the event"
+    /// );
+    /// # anyhow::Ok(()) });
+    /// ```
+    pub fn body_matches_partial_json(self, body: Value) -> Self {
+        Self { mock: self.mock.and(body_partial_json(body)), ..self }
+    }
+
+    /// Ensures that the send endpoint request uses a specific event type.
+    ///
+    /// Note: works with *any* room.
+    ///
+    /// # Examples
+    ///
+    /// see also [`MatrixMockServer::mock_room_send`] for more context.
+    ///
+    /// ```
+    /// # tokio_test::block_on(async {
+    /// use matrix_sdk::{
+    ///     ruma::{
+    ///         event_id,
+    ///         events::room::{
+    ///             create::RoomCreateEventContent, power_levels::RoomPowerLevelsEventContent,
+    ///         },
+    ///         events::StateEventType,
+    ///         room_id,
+    ///     },
+    ///     test_utils::mocks::MatrixMockServer,
+    /// };
+    ///
+    /// let mock_server = MatrixMockServer::new().await;
+    /// let client = mock_server.client_builder().build().await;
+    ///
+    /// mock_server.mock_room_state_encryption().plain().mount().await;
+    ///
+    /// let room = mock_server.sync_joined_room(&client, room_id!("!room_id:localhost")).await;
+    ///
+    /// let event_id = event_id!("$some_id");
+    ///
+    /// mock_server
+    ///     .mock_room_send_state()
+    ///     .for_type(StateEventType::RoomPowerLevels)
+    ///     .ok(event_id)
+    ///     .expect(1)
+    ///     .mount()
+    ///     .await;
+    ///
+    /// let response_not_mocked = room.send_state_event(RoomCreateEventContent::new_v11()).await;
+    /// // The `m.room.reaction` event type should not be mocked by the server.
+    /// assert!(response_not_mocked.is_err());
+    ///
+    /// let response = room.send_state_event(RoomPowerLevelsEventContent::new()).await?;
+    /// // The `m.room.message` event type should be mocked by the server.
+    /// assert_eq!(
+    ///     event_id, response.event_id,
+    ///     "The event ID we mocked should match the one we received when we sent the event"
+    /// );
+    ///
+    /// # anyhow::Ok(()) });
+    /// ```
+    pub fn for_type(self, event_type: StateEventType) -> Self {
+        let endpoint = RoomSendStateEndpoint { event_type: Some(event_type), ..self.endpoint };
+        let matcher = path_regex(Self::generate_path_regex(&endpoint));
+        Self { endpoint, mock: self.mock.and(matcher), ..self }
```

Comment:
I think here it would be better to generate the path regex at the end, just before building the final `MatrixMock`.

Reply:
Where would that be?
There are numerous places we build the final MatrixMock. Do you want to implement all the methods from `MockEndpoint<'a, T>` in 
`MockEndpoint<'a, RoomSendStateEndpoint>`. What if someone adds a new method to the generic version that can build a matrix mock but forgets to update this version to build the regex?

Reply:
You can take a look at `RoomEventEndpoint` as an example: it stores an optional `room`, set to `None` first: 

https://github.com/matrix-org/matrix-rust-sdk/blob/80a48f53ad19a2aae050737a6e8dbb9fe502acb3/crates/matrix-sdk/src/test_utils/mocks.rs#L431

Then, we have a builder method to optionally set it:

https://github.com/matrix-org/matrix-rust-sdk/blob/80a48f53ad19a2aae050737a6e8dbb9fe502acb3/crates/matrix-sdk/src/test_utils/mocks.rs#L995-L998

Then, we create the final path when we want to get the final `MatrixMock`:

https://github.com/matrix-org/matrix-rust-sdk/blob/80a48f53ad19a2aae050737a6e8dbb9fe502acb3/crates/matrix-sdk/src/test_utils/mocks.rs#L1009-L1023

If this code were to be duplicated in multiple methods, then it could be factored out as an helper method, like `generate_path_regexp` does, in fact.

Does it make sense now?